Fix dependency on redis-cli and redis-server to pass tests

avifenesh opened the pull request:
<!--
Thanks for contributing to Valkey GLIDE!

Please make sure you are aware of our contributing guidelines [available
here](https://github.com/valkey-io/valkey-glide/blob/main/CONTRIBUTING.md)

-->

Trying to work with Fedora which has strict rules about not OSS packages.
The Redis symlink doesn't exist in the distro because of it. When running tests, they fail, since there are many places we depend on the symlinks to exist.

This pr decouple and allow to developer to have just valkey on their system.

This pr also introduce the way to download Valkey using the package managers of the different distros, since those are not explained yet on Valkey sites.

### Issue link

This Pull Request is linked to issue #3011 

### Checklist

Before submitting the PR make sure the following are checked:

-   [x] This Pull Request is related to one issue.
-   [x] Commit message has a detailed description of what changed and why.
-   [x] Tests are added or updated.
-   [ ] CHANGELOG.md and documentation files are updated.
-   [x] Destination branch is correct - main or release
-   [x] Create merge commit if merging release branch into main, squash otherwise.

Yury-Fridlyand commented:
I propose to split this into 3 PRs, because it contains independent features - doc update, node client fixes and core fixes

avifenesh commented:
> I propose to split this into 3 PRs, because it contains independent features - doc update, node client fixes and core fixes

I'll split into two, doc and tests. The core one is just one change and naming, not sure if it needs a PR for itself.
It actually was a small fix, since I changed to fedora for a bit, discovered that yum already has it and that we are breaking on the valkey installed on fedora because of the strict rules. Then checked I thought, well, if fedora has it, probably there's another. And apparently all the main ones does so, added to docs.

BTW, on top of mac, Ubuntu and for node, also alpine, glide is compatible with Fedora and Mint, I checked it, and most likely all the CentOS/RHEL family and all the successors of Debian.
Just Arch is in question.

shohamazon commented on csharp/DEVELOPER.md:
```diff
@@ -1,6 +1,6 @@
 # Developer Guide
 
-This document describes how to set up your development environment to build and test the GLIDE for Redis C# wrapper.
+This document describes how to set up your development environment to build and test the GLIDE for valkey C# wrapper.
```
```suggestion
This document describes how to set up your development environment to build and test the Valkey GLIDE C# wrapper.
```

shohamazon commented on csharp/DEVELOPER.md:
```diff
@@ -11,7 +11,7 @@ The C# client contains the following parts:
 2. C# part of the client located in `lib`; it translates Rust async API into .Net async API.
 3. Integration tests for the C# client located in `tests` directory.
-4. A dedicated benchmarking tool designed to evaluate and compare the performance of GLIDE for Redis and other .Net clients. It is located in `<repo root>/benchmarks/csharp`.
+4. A dedicated benchmarking tool designed to evaluate and compare the performance of GLIDE for valkey and other .Net clients. It is located in `<repo root>/benchmarks/csharp`.
```
```suggestion
4. A dedicated benchmarking tool designed to evaluate and compare the performance of Valkey GLIDE and other .Net clients. It is located in `<repo root>/benchmarks/csharp`.
```

shohamazon commented on node/DEVELOPER.md:
````diff
@@ -56,6 +56,27 @@ curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
 ```
 
+**Valkey Server and CLI**
+For testing purposes, you need to install the Valkey server and CLI. You can follow the instructions [here](https://valkey.io/topics/installation/).
````
why not make it similar to go and c#?
```suggestion
See the [Valkey installation guide](https://valkey.io/topics/installation/) to install the Valkey server and CLI.
```

shohamazon commented on python/DEVELOPER.md:
```diff
@@ -124,7 +124,8 @@ maturin develop --release --strip
 
-Ensure that you have installed `redis-server` or `valkey-server` along with `redis-cli` or `valkey-cli` on your host. You can find the Redis installation guide at the following link: [Redis Installation Guide](https://redis.io/docs/install/install-redis/install-redis-on-linux/). You can get Valkey from the following link: [Valkey Download](https://valkey.io/download/).
+Ensure you have installed `valkey-server` and `valkey-cli` on your host (or `redis-server` and `redis-cli`). 
+You can find the Valkey installation guide at the following link: [Valkey Installation Guide](https://valkey.io/topics/installation/).
```
same, why not
```suggestion
See the [Valkey installation guide](https://valkey.io/topics/installation/) to install the Valkey server and CLI.
```

shohamazon commented on node/tests/TestUtilities.ts:
```diff
@@ -194,16 +195,30 @@ export async function GetAndSetRandomValue(client: Client) {
+                    );
+                }
+
+                const cli_1 = redisCliAvailable ? "valkey-cli" : "redis-cli";
```
minor: maybe serverCLI

shohamazon replied:
maybe instead of checking availability on both each time and then checking if both exists etc, just create a function which returns valkey or redis and that will do all the work

shohamazon replied:
and shouldnt it be
```suggestion
                const cli_1 = valkeyCliAvailable ? "valkey-cli" : "redis-cli";
```

shohamazon commented on node/tests/TestUtilities.ts:
```diff
@@ -2106,3 +2121,65 @@ export async function getServerVersion(
+}
+
+/**
+ * Starts a Redis-compatible server on the specified port
```
```suggestion
 * Starts a Valkey/Redis-compatible server on the specified port
```

shohamazon commented on node/.ort.yml:
```diff
@@ -8,11 +8,6 @@ excludes:
-        - message: "NPM failed to resolve dependencies for path 'hybrid-node-tests/ecmascript-test/package.json': FileNotFoundException: .*valkey-glide/package.json.*"
-          reason: "BUILD_TOOL_ISSUE"
-          comment: "The 'hybrid-node-tests' are not built, so there is no 'package.json' in the expected location. The 'valkey-glide' package has already been scanned for licenses, so there's no need to scan it again."
+        - pattern: "hybrid-node-tests/**"
```
what are we doing here?

avifenesh replied:
The hybrid tests are dependent on glide files. When getting build we getting the same package (glide) in both projects, and the ort is getting crazy. Those are just testing files, which the only package they use is find free port, an MIT licensed which hasn't had any release for 2 years. So it doesn't need to go through ort, and it is constantly causing issue, so decided to skip it.

shohamazon commented on node/hybrid-node-tests/commonjs-test/commonjs-test.cjs:
```diff
@@ -1,57 +1,76 @@
+    }
+
+    const server = valkeyServerAvailable ? "valkey-server" : "redis-server";
+    console.log(server);
```
```suggestion
    console.log("Server is: " + server);
```

shohamazon commented on node/hybrid-node-tests/commonjs-test/commonjs-test.cjs:
```diff
@@ -1,57 +1,76 @@
-        console.error("Error occurred while finding a free port:", error);
     });
+
+    const client = await GlideClient.createClient({
```
why is it no longer AsyncClient? is AsyncClient an old glide name?

avifenesh replied:
I have no idea why it was there from the beginning, its pointing to the glide core

shohamazon commented on node/hybrid-node-tests/ecmascript-test/ecmascript-test.mjs:
```diff
@@ -1,50 +1,73 @@
+
+function checkCommandAvailability(command) {
+    return new Promise((resolve) => {
+        execFile(command, ["--version"], (error) => {
```
why in some places you check like this and in other with which?

avifenesh replied:
hmmm idk, i checked it like this everywhere, and it crashed, just here not. ill use which.

shohamazon commented on node/tests/AsyncClient.test.ts:
```diff
@@ -4,44 +4,39 @@
     runCommonTests({
         init: async () => {
-            const client = await AsyncClient.CreateConnection(
+            const client = AsyncClient.CreateConnection(
```
why?

avifenesh replied:
because it is not async call. whosever wrote it just added await for no reason.

shohamazon commented on node/tests/TestUtilities.ts:
```diff
@@ -2106,3 +2121,65 @@ export async function getServerVersion(
+            checkCliAvailability("valkey-server"),
+            checkCliAvailability("redis-server"),
+        ]).then(([valkeyExists, redisExists]) => {
+            if (!valkeyExists && !redisExists) {
```
again, you check that both return an answer, cant it be like this?
https://github.com/valkey-io/valkey-glide/blob/6a98816ad8efeb12cee01362543f1264bbffa0e3/utils/cluster_manager.py#L346

avifenesh replied:
I actually think this one is faster - i check both together, instead of loop one by one.
The problem is that i check both, or the codding style?

shohamazon replied:
Just the coding style, I see that you have multiple checks of:
```
valkeyCli =  check(valkey)
redisCli = check(redis)

if not redis and not valkey -> raise an error

server = valkeyCli? "valkey-cli" : "redis-cli"
```

this block is just repeating itself in many places, even twice in [node/tests/TestUtilities.ts](https://github.com/valkey-io/valkey-glide/pull/3014/files#diff-6cc2fb6a9baaaa3fdf3e95bc64aa27155ae36ecf17b6a9f94797a32446ab103a)

I think it would be easier to have a function that returns the available server

avifenesh commented on eslint.config.mjs:
```diff
@@ -2,21 +2,21 @@
 import eslint from "@eslint/js";
 import prettierConfig from "eslint-config-prettier";
 import tseslint from "typescript-eslint";
-import jsdoc from "eslint-plugin-jsdoc";
```
not in use

avifenesh commented on node/hybrid-node-tests/commonjs-test/package.json:
```diff
@@ -14,13 +14,10 @@
-        "child_process": "^1.0.2",
-        "find-free-port": "^2.0.0",
-        "valkey-glide": "file:../../../node/build-ts/cjs"
+        "@valkey/valkey-glide": "file:../..",
```
Avoiding collisions. Used npm link

avifenesh commented on node/npm/glide/package.json:
```diff
@@ -33,16 +33,11 @@
-        "@types/node": "^18.11.18",
-        "@typescript-eslint/eslint-plugin": "^5.48.0",
-        "@typescript-eslint/parser": "^5.48.0",
+        "@types/node": "^22.13.1",
```
Non-relevant here
